### PR TITLE
vault: recover from standby losing etcd lease (#3031)

### PR DESCRIPTION
**Pull request by bhiggins**
This change makes these errors transient instead of permanent:

[ERROR] core: failed to acquire lock: error=etcdserver: requested lease not found

After this change, there can still be one of these errors when a
standby vault that lost its lease tries to become leader, but on the
next lock acquisition attempt a new session will be created. With this
new session, the standby will be able to become the leader.

### CODE REVIEW

**Comment by bhiggins** [comment]
Fixes #3031

File: physical/etcd/etcd3.go [hunk 1]
```diff
@@ -267,6 +267,25 @@ func (c *EtcdLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
...
+	case _, ok := <-c.etcdSession.Done():
+		if !ok {
+			// The session's done channel is closed, so the session is over
+			needNewSession = true
+		}
+	default:
+	}
+
+	if needNewSession {
+		session, err := concurrency.NewSession(c.etcd, concurrency.WithTTL(etcd3LockTimeoutInSeconds))
```
Comment (xiang90) [inline physical/etcd/etcd3.go]: move the code to line 273 directly?

**Comment by xiang90** [comment]
@bhiggins the fix looks right. can you simplify the code? have you tested it manually? if yes, then it looks good to me.

**Comment by xiang90** [comment]
@bhiggins thanks!

File: physical/etcd/etcd3.go (hunk 1 above)
Comment (bhiggins) [reply physical/etcd/etcd3.go]: Done.

**Comment by bhiggins** [comment]
Simplified code as requested.

Yes, I tested this patch against 0.7.3 by having three vaults running in HA config. I manually deleted the standby's leases out of etcd, and then killed the leader. After that, as described in the commit message, the "lease not found" error occurs once but then on the next lock attempt a new session is created and the standby becomes the leader.

**Comment by xiang90** [comment]
lgtm. defer to @jefferai

**Comment by jefferai** [comment]
Hah, it's the other way, when it comes to etcd3 I defer to @xiang90 :-D

Merging.

**Pull request merged**